File: backend/internal/api/resource_email.go
package api

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/mlorentedev/mlorente-backend/internal/models"
	"github.com/mlorentedev/mlorente-backend/internal/services"
	"github.com/mlorentedev/mlorente-backend/pkg/logger"
)

// ResourceEmailHandler maneja las solicitudes para enviar email con recurso
func ResourceEmailHandler(c *gin.Context) {
	var request struct {
		Email         string `json:"email" binding:"required"`
		ResourceID    string `json:"resourceId" binding:"required"`
		ResourceTitle string `json:"resourceTitle"`
		ResourceLink  string `json:"resourceLink" binding:"required"`
	}

	if err := c.ShouldBindJSON(&request); err != nil {
		logger.LogFunction("error", "Error in resource email endpoint", err.Error())
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"message": "Datos incompletos para completar la operación.",
		})
		return
	}

	// Validar email
	if !services.IsValidEmailFormat(request.Email) {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"message": "Correo electrónico inválido.",
		})
		return
	}

	// Enviar email con recurso
	resourceTitle := request.ResourceTitle
	if resourceTitle == "" {
		resourceTitle = services.GenerateResourceTitle(request.ResourceID, "")
	}

	emailSent, err := services.SendResourceEmail(models.ResourceEmailOptions{
		Email:         request.Email,
		ResourceID:    request.ResourceID,
		ResourceTitle: resourceTitle,
		ResourceLink:  request.ResourceLink,
	})

	if err != nil {
		logger.LogFunction("error", "Error sending resource email", err.Error())
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"message": "Error interno del servidor.",
		})
		return
	}

	if emailSent {
		logger.LogFunction("info", "Resource email sent successfully", map[string]string{"email": request.Email, "resourceId": request.ResourceID})
		c.JSON(http.StatusOK, gin.H{
			"success": true,
			"message": "Email enviado correctamente.",
		})
	} else {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"message": "Error interno del servidor.",
		})
	}
}

File: src/pages/api/resource-email.ts
// src/api/resource-email.ts

import type { APIRoute } from 'astro';
import { FRONTEND_MESSAGES } from '../../config/constants';
import { generateResourceTitle } from '../../domain/business';
import { validateEmailConfiguration, sendResourceEmail } from '../../services/email.service';
import { logFunction } from '../../utils/logging';
import { validateEmail } from '../../utils/validation';

export const prerender = false;

/**
 * Endpoint POST para enviar email con recurso
 */
export const POST: APIRoute = async ({ request }) => {
  try {
    // Validar configuración de email
    if (!validateEmailConfiguration()) {
      return new Response(
        JSON.stringify({
          success: false,
          message: FRONTEND_MESSAGES.ERRORS.EMAIL_CONFIG_ERROR,
        }),
        { status: 500 }
      );
    }

    const { email, resourceId, resourceTitle, resourceLink } = await request.json();

    // Validar datos requeridos
    if (!email || !resourceId || !resourceLink) {
      return new Response(
        JSON.stringify({
          success: false,
          message: FRONTEND_MESSAGES.ERRORS.INCOMPLETE_DATA,
        }),
        { status: 400 }
      );
    }

    // Validar email
    if (!validateEmail(email)) {
      return new Response(
        JSON.stringify({
          success: false,
          message: FRONTEND_MESSAGES.ERRORS.INVALID_EMAIL,
        }),
        { status: 400 }
      );
    }

    // Enviar email con recurso
    const emailSent = await sendResourceEmail({
      email,
      resourceId,
      resourceTitle: resourceTitle || generateResourceTitle(resourceId),
      resourceLink,
    });

    if (emailSent) {
      logFunction('info', 'Resource email sent successfully', { email, resourceId });
      return new Response(
        JSON.stringify({
          success: true,
          message: FRONTEND_MESSAGES.SUCCESS.EMAIL_SENT,
        }),
        { status: 200 }
      );
    } else {
      return new Response(
        JSON.stringify({
          success: false,
          message: FRONTEND_MESSAGES.ERRORS.SERVER_ERROR,
        }),
        { status: 500 }
      );
    }
  } catch (error) {
    logFunction('error', 'Error in resource email endpoint', error);
    return new Response(
      JSON.stringify({
        success: false,
        message: FRONTEND_MESSAGES.ERRORS.SERVER_ERROR,
      }),
      { status: 500 }
    );
  }
};
